keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -453,6 +449,18 @@ only applies enforcement to ephemeral containers (defaults to the allow policy).
 
 [custom security contexts]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/277-ephemeral-containers#configurable-security-policy
 
+One way this could be handled under the current model is:
+1. Exempt a special username (not one that can be authenticated directly) from policy enforcement,
+   e.g. `ops:privileged-debugger`
+2. Grant the special user permission to ONLY operate on the ephemeral containers subresource (it is
+   critical that they cannot create or update pods directly).
+3. Grant (real) users that should have privileged debug capability the ability to impersonate the
+   exempt user.
+
+We could consider ways to streamline the user experience of this, for instance adding a special RBAC
+binding that exempts users when operating on the ephemeral containers subresource (e.g. an
+`escalate-privilege` verb on the ephemeral containers subresource).
+
```

Comment:
Not specifically to this line, but the ephemeral container issue in general:

All of this is quite tricky. I'd like for my users to be able to debug their own pods. I'd prefer the pods not to have CAP_SYS_PTRACE when not needed.

I don't want to have to name every user that needs special privilege to debug their own pods. And it should not be granted to users across namespaces. So RBAC really needs to be at play. If your a ClusterRole Admin or Maybe Debugger or something ClusterRole, then you can use the feature.

Also, most of our permissions are not based on username, but by group. Our usernames often are so ephemeral, that they sometimes even change during a users lifetime. K8s usually doesn't care when you use groups for perms. So again, RBAC helps with that.

Reply:
Yeah, that all makes sense. It's a good point about wanting to grant the debug permissions at namespace granularity, which breaks with the impersonation approach, but is solved by the special RBAC verb. I think this is probably not something we're going to solve for the alpha implementation, but we should aim for a solution around the beta timeline (depending on ephemeral containers status).

Reply:
Maybe short term, as a stop gap, a namespace being flagged as debugable slides in CAP_SYS_PTRACE to all pods in that namespace. Then ephemeral debug containers work as expected api wise? Then it can be hardened from there later.

Reply:
if the baseline policy disallows CAP_SYS_PTRACE, but you want arbitrary users to be able to add that to an ephemeral container on demand, it doesn't sound like those namespaces can use the baseline policy.

Reply:
> It's a good point about wanting to grant the debug permissions at namespace granularity, which breaks with the impersonation approach, but is solved by the special RBAC verb.

Write access to the `pods/ephemeralcontainers` endpoint can be granted per namespace with RBAC, and is not included by default in the `admin` or `edit` RBAC roles. That seems sufficient to control access to debugging.

As mentioned above, I'm generally opposed to the idea of letting ephemeral containers exceed the namespace policy level. That would mean selecting a policy level for the namespace that permits the capabilities you want your users to be able to make use of in their ephemeral containers.